--- a/src/ortc.cpp	2021-06-14 23:47:03.000000000 +0900
+++ b/src/ortc.cpp	2021-07-01 20:27:19.925000000 +0900
@@ -4,6 +4,8 @@
 #include "Logger.hpp"
 #include "MediaSoupClientErrors.hpp"
 #include <media/base/h264_profile_level_id.h>
+#include <media/base/sdp_video_format_utils.h>
+#include <pc/webrtc_sdp.h>
 #include <algorithm> // std::find_if
 #include <regex>
 #include <stdexcept>
@@ -1600,8 +1602,8 @@ static bool matchCodecs(json& aCodec, js
 		// If strict matching check profile-level-id.
 		if (strict)
 		{
-			webrtc::H264::CodecParameterMap aParameters;
-			webrtc::H264::CodecParameterMap bParameters;
+			cricket::CodecParameterMap aParameters;
+			cricket::CodecParameterMap bParameters;
 
 			aParameters["level-asymmetry-allowed"] = std::to_string(getH264LevelAssimetryAllowed(aCodec));
 			aParameters["packetization-mode"]      = std::to_string(aPacketizationMode);
@@ -1613,11 +1615,11 @@ static bool matchCodecs(json& aCodec, js
 			if (!webrtc::H264::IsSameH264Profile(aParameters, bParameters))
 				return false;
 
-			webrtc::H264::CodecParameterMap newParameters;
+			cricket::CodecParameterMap newParameters;
 
 			try
 			{
-				webrtc::H264::GenerateProfileLevelIdForAnswer(aParameters, bParameters, &newParameters);
+				webrtc::H264GenerateProfileLevelIdForAnswer(aParameters, bParameters, &newParameters);
 			}
 			catch (std::runtime_error)
 			{
