diff --git a/include/Handler.hpp b/include/Handler.hpp
index e306975..f2bb5f3 100644
--- a/include/Handler.hpp
+++ b/include/Handler.hpp
@@ -101,7 +101,7 @@ namespace mediasoupclient
 
 	public:
 		SendResult Send(
-		  webrtc::MediaStreamTrackInterface* track,
+		  rtc::scoped_refptr<webrtc::MediaStreamTrackInterface> track,
 		  std::vector<webrtc::RtpEncodingParameters>* encodings,
 		  const nlohmann::json* codecOptions,
 		  const nlohmann::json* codec);
diff --git a/include/PeerConnection.hpp b/include/PeerConnection.hpp
index 93b8147..c34a317 100644
--- a/include/PeerConnection.hpp
+++ b/include/PeerConnection.hpp
@@ -126,7 +126,7 @@ namespace mediasoupclient
 		  rtc::scoped_refptr<webrtc::MediaStreamTrackInterface> track,
 		  webrtc::RtpTransceiverInit rtpTransceiverInit);
 		std::vector<rtc::scoped_refptr<webrtc::RtpSenderInterface>> GetSenders();
-		bool RemoveTrack(webrtc::RtpSenderInterface* sender);
+		webrtc::RTCError RemoveTrack(rtc::scoped_refptr<webrtc::RtpSenderInterface> sender);
 		nlohmann::json GetStats();
 		nlohmann::json GetStats(rtc::scoped_refptr<webrtc::RtpSenderInterface> selector);
 		nlohmann::json GetStats(rtc::scoped_refptr<webrtc::RtpReceiverInterface> selector);
diff --git a/include/Transport.hpp b/include/Transport.hpp
index 2304fcc..47a88d7 100644
--- a/include/Transport.hpp
+++ b/include/Transport.hpp
@@ -129,7 +129,7 @@ namespace mediasoupclient
 	public:
 		Producer* Produce(
 		  Producer::Listener* producerListener,
-		  webrtc::MediaStreamTrackInterface* track,
+		  rtc::scoped_refptr<webrtc::MediaStreamTrackInterface> track,
 		  const std::vector<webrtc::RtpEncodingParameters>* encodings,
 		  const nlohmann::json* codecOptions,
 		  const nlohmann::json* codec,
diff --git a/src/Handler.cpp b/src/Handler.cpp
index 6120657..e8524c3 100644
--- a/src/Handler.cpp
+++ b/src/Handler.cpp
@@ -168,7 +168,7 @@ namespace mediasoupclient
 	};
 
 	SendHandler::SendResult SendHandler::Send(
-	  webrtc::MediaStreamTrackInterface* track,
+	  rtc::scoped_refptr<webrtc::MediaStreamTrackInterface> track,
 	  std::vector<webrtc::RtpEncodingParameters>* encodings,
 	  const json* codecOptions,
 	  const json* codec)
diff --git a/src/PeerConnection.cpp b/src/PeerConnection.cpp
index 849db1f..f608da9 100644
--- a/src/PeerConnection.cpp
+++ b/src/PeerConnection.cpp
@@ -317,11 +317,11 @@ namespace mediasoupclient
 		return this->pc->GetSenders();
 	}
 
-	bool PeerConnection::RemoveTrack(webrtc::RtpSenderInterface* sender)
+	webrtc::RTCError PeerConnection::RemoveTrack(rtc::scoped_refptr<webrtc::RtpSenderInterface> sender)
 	{
 		MSC_TRACE();
 
-		return this->pc->RemoveTrack(sender);
+		return this->pc->RemoveTrackOrError(sender);
 	}
 
 	json PeerConnection::GetStats()
diff --git a/src/Transport.cpp b/src/Transport.cpp
index fb57be5..c9952d7 100644
--- a/src/Transport.cpp
+++ b/src/Transport.cpp
@@ -174,7 +174,7 @@ namespace mediasoupclient
 	 */
 	Producer* SendTransport::Produce(
 	  Producer::Listener* producerListener,
-	  webrtc::MediaStreamTrackInterface* track,
+	  rtc::scoped_refptr<webrtc::MediaStreamTrackInterface> track,
 	  const std::vector<webrtc::RtpEncodingParameters>* encodings,
 	  const json* codecOptions,
 	  const json* codec,
